pkg/controllers/subnetbinding/subnetbinding_controller.go

```diff
@@ -203,52 +203,110 @@ func (r *Reconciler) listBindingMapIDsFromCRs(ctx context.Context) (sets.Set[str
 }
 
+func getVpcPath(subnetPath string) (string, *errorWithRetry) {
+	info, err := servicecommon.ParseVPCResourcePath(subnetPath)
+	if err != nil {
+		return "", &errorWithRetry{
+			message: fmt.Sprintf("Invalid Subnet path %s", subnetPath),
+			retry:   false,
+			error:   fmt.Errorf("failed to parse Subnet path %s", subnetPath),
+		}
+	}
+	return info.GetVPCPath(), nil
+}
+
 // validateDependency validates the following conditions:
 //  1. the dependent Subnet/SubnetSet is not realized. In this case, a not-retry error is returned, and the
 //     Subnet/SubnetSet readiness update will actively trigger a requeue event
 //  2. the associated Subnet is already used as a target Subnet in another SubnetConnectionBindingMap CR, or the target
 //     Subnet already has associated SubnetConnectionBindingMap CR. In this case, a retry error is returned.
+//  3. the target Subnet is a pre-created Subnet. In this case, a not-retry error is returned.
+//  4. the associated Subnet is a pre-created Subnet in a VPC different from the target Subnet Namespace VPC
+//     In this case, not-retry error is returned.
 func (r *Reconciler) validateDependency(ctx context.Context, bindingMap *v1alpha1.SubnetConnectionBindingMap) (*model.VpcSubnet, []*model.VpcSubnet, *errorWithRetry) {
-	childSubnets, err := r.validateVpcSubnetsBySubnetCR(ctx, bindingMap.Namespace, bindingMap.Spec.SubnetName, false)
+	childSubnets, childSubnetCR, err := r.validateVpcSubnetsBySubnetCR(ctx, bindingMap.Namespace, bindingMap.Spec.SubnetName, false)
 	if err != nil {
 		return nil, nil, err
 	}
 	childSubnet := childSubnets[0]
 
+	var parentSubnets []*model.VpcSubnet
 	if bindingMap.Spec.TargetSubnetName != "" {
-		parentSubnets, err := r.validateVpcSubnetsBySubnetCR(ctx, bindingMap.Namespace, bindingMap.Spec.TargetSubnetName, true)
+		var parentSubnetCR *v1alpha1.Subnet
+		parentSubnets, parentSubnetCR, err = r.validateVpcSubnetsBySubnetCR(ctx, bindingMap.Namespace, bindingMap.Spec.TargetSubnetName, true)
+		if err != nil {
+			return nil, nil, err
+		}
+		// Check if the target Subnet is pre-created Subnet
+		if parentSubnetCR.Status.Shared {
+			return nil, nil, &errorWithRetry{
+				message: fmt.Sprintf("Target Subnet %s/%s is a pre-created Subnet", bindingMap.Namespace, bindingMap.Spec.TargetSubnetName),
+				error:   fmt.Errorf("pre-created Subnet %s/%s cannot be a target Subnet", bindingMap.Namespace, bindingMap.Spec.TargetSubnetName),
+				retry:   false,
+			}
+		}
+	} else {
+		parentSubnets, err = r.validateVpcSubnetsBySubnetSetCR(ctx, bindingMap.Namespace, bindingMap.Spec.TargetSubnetSetName)
 		if err != nil {
 			return nil, nil, err
 		}
-		return childSubnet, parentSubnets, nil
 	}
 
-	parentSubnets, err := r.validateVpcSubnetsBySubnetSetCR(ctx, bindingMap.Namespace, bindingMap.Spec.TargetSubnetSetName)
-	if err != nil {
-		return nil, nil, err
+	// If child Subnet is a pre-created Subnet, check if it is in the same vpc as parent Subnet
+	if childSubnetCR.Status.Shared {
```

Comment:
could we just check the "nsx.vmware.com/associated_resource" annotation to check if the parent and child Subnet are in the same VPC? we do not need to get the real NSX Subnet.

Reply:
We always need to get child subnet path for bindingmap creation. But as only the path is required, we can save the API call to get the real NSX Subnet by changing the return value of those bindingmap functions. I updated the code, please take another look, thanks!

Reply:
make sense.